<table>
        <td>
        <td><img src="https://raw.githubusercontent.com/Fabian830348/Bases_Datos/refs/heads/master/Logo_EICT_horizontal_ESPANOL%20(1).png" width="400"/></td>
        <td>&nbsp;</td>
        <td>
            <h1 style="font-size:200%;color:blue;text-align:center">    <FONT COLOR="blue"> Exploración de Datos  </p> EDA </p> Python   </FONT>         </h1></td>         
        <td>
            <tp><p style="font-size:99%;text-align:center">Machine Learning </p></tp>
            <tp><p style="font-size:115%;text-align:center">Diplomado 2025-2</p></tp>
            <tp><p style="font-size:115%;text-align:center">Prof. Fabián Sánchez</p></tp>
        </td>
    </tr>
</table>

In [ ]:
%%html
<center><marquee style='width: 60%; color: red'><b>✌ ¡Hola a todos y a todas! Un saludo... Bienvenidos a la introducción a Python-Colab✌ </b></marquee><center>

# <FONT SIZE=5 COLOR="green"> 1. Python </FONT>

Python es un lenguaje de programación de propósito general, usado ampliamente para

*  Desarrollo de aplicaciones
*  Desarrollo web
*  Web Scraping
*  **Ciencia de Datos**
*  **Visualización**
*  Aplicaciones a los negocios
*  **Machine Learning**
*  **Deep Learning**

Python cuenta con librerías especializadas dependiendo del tipo de tarea. Algunas de ellas son:

<center><img src="https://github.com/Fabian830348/cursos/blob/master/librerias.png?raw=true" alt="centered image" width="550" height="350"></center> <center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: https://www.actumlogos.com/python-para-ia.html  </FONT> <figcaption></center>

1.   **Pandas:** manipulación de datos-estadísticas-data.frames
2.   **Matplotlib**: visualización-gráficas
3.   **numpy**: operaciones matemáticas
4.   **spicy**: realizar cálculo simbólico
5.   **scikit-learn**: ejecutar algoritmos de machine learning
6.   **NTKL**: Procesamiento de Lenguaje Natural.
7.   **seaborn**: creación de gráficos estadísticos
8.   **Datetime**: para manipulación de fechas



---



<center><FONT SIZE=5 COLOR="purple"> EXPLORACIÓN DE DATOS EN PYTHON</FONT></center>

<center><img src="https://miro.medium.com/v2/resize:fit:720/format:webp/0*fto5mNElXLZ8TCZr.png" alt="centered image" width="500" height="250"></center>

+ Librería de Python especializada manipulación y análisis de estructuras de datos
+ Permite leer datos en diferentes formatos: csv, excel, bases de datos SQL.
+ Realiza operaciones rápidas y eficientes sobre conjuntos de datos
+ Permite trabajar series de tiempo
+ *import pandas as pd*
+ [Guía Pandas](https://aprendeconalf.es/docencia/python/manual/pandas/)



<center><FONT SIZE=5 COLOR="blue"> DATAFRAMES </FONT></center>

Los tipos de datos que vamos a trabajar se denominan *dataframes*

<center><img src="https://github.com/Fabian830348/cursos/blob/master/dataframe.png?raw=true" alt="centered image" width="600" height="300"></center>

Observe que un *dataframe* puede tener diferentes tipos de variables

<center><img src="https://github.com/Fabian830348/cursos/blob/master/tiposvariables.png?raw=true" alt="centered image" width="600" height="400"></center>





---



<center><FONT SIZE=5 COLOR="PURPLE"> PROCESO DE EXPLORACIÓN DE LOS DATOS </FONT></center>

<br>

<center><img src="https://github.com/Fabian830348/cursos/blob/master/imagen_eda.png?raw=true" alt="centered image" width="550" height="350"></center><center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: Elaboración propia </FONT> <figcaption></center>

En el gráfico anterior se muestran algunas etapas fundamentales:

1.	**Preguntas para responder**. Es importante tener claro para qué queremos hacer el procesamiento y análisis de datos. Es decir, qué buscamos responder.

2.	**Generalidades del conjunto de datos**.  Acá buscamos reconocer la dimensión, datos faltantes, y en general, hacer un resumen del conjunto de datos.

3.	**Tipos de Variables**. Clasificar las variables en cualitativas y cuantitativas y examinar como se presenta esta información. Por ejemplo, una variable puede estar codificada y verse como numérica, pero ser categórica.

4.	**Estadística Descriptiva**. Se busca resumir la información. Utilizar medidas de tendencia, de dispersión, de forma, de posición, etc., para entender mejor los datos. Por ejemplo, se identifican datos atípicos usando boxplot.

5.	**Visualización**.  Es una etapa fundamental en el análisis de datos. Visualizar la información y poder describirla gráficamente. En esta etapa se usan: diagramas de barras, pie, boxplot, histogramas, líneas de densidad, diagramas de dispersión, entre otros.

6.	**Relación entre variable**. Se exploran datos bivariados. Se examinan relaciones entre variables cualitativas (tablas de contingencia) y relaciones entre variables cuantitativas (diagramas de dispersión) se establecen posibles relaciones lineales.

7.	**Reporte de resultados**. Una vez se haga el proceso de exploración y análisis en Python, R-Studio u otro, se deben reportar los resultados y conclusiones de lo que se encontró en el conjunto de datos.

<center><FONT SIZE=5 COLOR="blue"> ¿Cómo importamos una librería para trabajar?</FONT></center>

In [ ]:
# librería para manejo de dataframes
import pandas as pd
# librería para hacer ciertos cálculos y funciones
import numpy  as np

# <FONT SIZE=5 COLOR="green"> 2. Carga de Datos en Python </FONT>

Para cargar datos en Google-Colab podemos usar varias alternativas, por ejemplo

- Desde el drive

- Repositorio externo como Github u otro.

Se pueden cargar datos en diferentes formatos, por ejemplo:

- $read\_csv$ : archivos separados por comas

- $read\_excel$: clásicas hojas de cálculo

- $read\_json$ : javascript object notation. (formato java)

## <FONT SIZE=4 COLOR="magenta"> **2.1 Desde el drive** </FONT>

Con la siguiente instrucción conectamos google-colab con Drive para poder cargar archivos desde este último.

```python
from google.colab import drive
drive.mount('/content/drive')
```



In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


Ya hemos importado la librería ***pandas*** previamente.

**Ejemplo 1**

Vamos a importar el conjunto de datos denominado $BBDD\_Hospitalización.xlsx$.

$$pd.read\_excel(''ruta'')$$

Esta es la ruta donde tengo los archivos en el DRIVE (para ver esta ruta revisar la carpetica que esta en la barra izquierda, luego entran al drive y buscan el archivo, le dan click a los tres puntos y donde dice copiar ruta)

In [ ]:
# así se cargan datos. El nombre de hospital yo lo asigné
hospital = pd.read_excel("/content/drive/MyDrive/1.2 SEMINARIO IA-SALUD/BBDD_Hospitalizacion.xlsx")

In [ ]:
# vamos a ver los datos
hospital

EDAD DIABETES HOSPITALIZACIÓN ULTIMO MES    PSA BIOPSIAS PREVIAS  \
0    53.0       NO                         NO   4.00               NO   
1    56.0       NO                         NO   7.70               NO   
2    57.0       NO                         NO  29.00               SI   
3    56.0       NO                         NO   7.00               NO   
4    55.0       NO                         NO  29.00               SI   
..    ...      ...                        ...    ...              ...   
565  69.0       NO                         NO   4.43              NaN   
566  67.0       NO                         NO   8.00               NO   
567  59.0       NO                         NO   9.00               SI   
568   NaN      NaN                        NaN    NaN              NaN   
569   NaN      NaN                        NaN    NaN              NaN   

    VOLUMEN PROSTATICO ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS  \
0                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
1                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
2                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
3                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
4                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
..                 ...                                     ...   
565                NaN          FLUOROQUINOLONA_AMINOGLICOSIDO   
566                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
567                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
568                NaN                                     NaN   
569                NaN                                     NaN   

     NUMERO DE MUESTRAS TOMADAS  CUP ENF. CRONICA PULMONAR OBSTRUCTIVA  \
0                          12.0   NO                                NO   
1                          12.0   NO                                NO   
2                          24.0   NO                                NO   
3                          12.0   NO                                NO   
4                          24.0   NO                                NO   
..                          ...  ...                               ...   
565                        20.0  NaN                                NO   
566                        12.0   NO                               NaN   
567                        12.0   NO                               NaN   
568                         NaN  NaN                               NaN   
569                         NaN  NaN                               NaN   

                       BIOPSIA  \
0                          NEG   
1                          NEG   
2                          NEG   
3                          NEG   
4                          NEG   
..                         ...   
565                        NEG   
566  ADENOCARCINOMA GLEASON 7    
567                        NEG   
568                        NaN   
569                        NaN   

    NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA  \
0                                                    1                          
1                                                    1                          
2                                                    1                          
3                                                    1                          
4                                                    1                          
..                                                 ...                          
565                                                 NO                          
566                                                 NO                          
567                                                 NO                          
568                                                NaN                          
569                                                NaN                          

    FIEBRE  ITU TIPO DE CULTIVO AGENTE AISLADO  \
0       SI   N

Vamos a importar otro conjunto de datos

**Ejemplo 2**

Vamos a importar el conjunto de datos denominado $Apartamentos.csv$.

$$pd.read\_csv(''ruta'')$$

In [ ]:
aptos = pd.read_csv("/content/drive/MyDrive/1.2 SEMINARIO IA-SALUD/Apartamentos.csv")

In [ ]:
aptos.head()

**Nota importante:**
Cuando cargamos datos desde el *DRIVE* las rutas de acceso de los conjuntos de datos son los correspondientes al propietario del notebook.

En ese orden de ideas, si vamos a ejecutar las líneas de código en otra cuenta debemos cambiar las rutas de acceso o de lo contrario, producirá un error.

Para sortear esta situación podemos cargar los datos desde un reporsitorio externo, por ejemplo, desde el ***GitHub*** o desde internet directamente.

## <FONT SIZE=4 COLOR="magenta"> **2.2 Repositorio Externo Github** </FONT>

- Ir al Github (en mi caso FABIAN830348)
https://github.com/Fabian830348

- Copiar la ruta de los datos y se pega acá.

In [ ]:
pd.read_csv("https://raw.githubusercontent.com/jbagnato/machine-learning/refs/heads/master/comprar_alquilar.csv")

In [ ]:
cancer = pd.read_csv("https://raw.githubusercontent.com/Fabian830348/Bases_Datos/master/cancer_clasificacion.csv" , sep = ";" )

In [ ]:
cancer.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0            1.184             2.776           3.001                1.471   
1            8.474             7.864         869.000                7.017   
2            1.096             1.599           1.974                1.279   
3            1.425             2.839           2.414                1.052   
4            1.003             1.328         198.000                1.043   

   ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
0  ...         25.38          17.33           184.60      2019.0   
1  ...         24.99          23.41           158.80      1956.0   
2  ...         23.57          25.53           152.50      1709.0   
3  ...         14.91          26.50            98.87       567.7   
4  ...         22.54          16.67           152.20      1575.0   

   smoothness_worst  compactness_worst  concavity_worst  concave points_worst  \
0             1.622              6.656            7.119                 2.654   
1             1.238              1.866            2.416               186.000   
2             1.444              4.245            4.504               243.000   
3             2.098              8.663            6.869                 2.575   
4             1.374            205.000            0.400                 1.625   

   symmetry_worst  fractal_dimension_worst  
0           4.601                    1.189  
1         275.000                    8.902  
2           3.613                    8.758  
3           6.638                  173.000  
4           2.364                    7.678  

[5 rows x 32 columns]

In [ ]:
homicidios = pd.read_csv("https://raw.githubusercontent.com/Fabian830348/Bases_Datos/refs/heads/master/Homicidios.csv")
homicidios

Fecha Departamento            Municipio      Día      Hora  \
0        1/1/2017     AMAZONAS         LETICIA (CT)  Domingo  15:05:00   
1        1/1/2017    ANTIOQUIA              BETANIA  Domingo  21:20:00   
2        1/1/2017    ANTIOQUIA           COPACABANA  Domingo   4:48:00   
3        1/1/2017    ANTIOQUIA              EBÉJICO  Domingo   5:00:00   
4        1/1/2017    ANTIOQUIA             EL BAGRE  Domingo   6:30:00   
...           ...          ...                  ...      ...       ...   
11960  12/31/2017        VALLE            CALI (CT)  Domingo   0:31:00   
11961  12/31/2017        VALLE              CARTAGO  Domingo  17:12:00   
11962  12/31/2017        VALLE          LA VICTORIA  Domingo  21:20:00   
11963  12/31/2017      VICHADA  PUERTO CARREÑO (CT)  Domingo   2:00:00   
11964  12/31/2017      VICHADA  PUERTO CARREÑO (CT)  Domingo   2:00:00   

                  Barrio    Zona         Clase de sitio  \
0           KILOMETRO 11   RURAL              CARRETERA   
1              LA CUMBRE  URBANA          VIAS PUBLICAS   
2                MACHADO  URBANA  DENTRO DE LA VIVIENDA   
3                  SAGUA   RURAL     FINCAS Y SIMILARES   
4             LAS BRISAS  URBANA          VIAS PUBLICAS   
...                  ...     ...                    ...   
11960  EL POBLADO II E13  URBANA          VIAS PUBLICAS   
11961    URB. BELLAVISTA  URBANA          VIAS PUBLICAS   
11962             CENTRO  URBANA          VIAS PUBLICAS   
11963           ACEITICO   RURAL                   ISLA   
11964           ACEITICO   RURAL                   ISLA   

                     Arma empleada         Móvil Agresor  \
0      ARMA BLANCA / CORTOPUNZANTE  PASAJERO MOTOCICLETA   
1                    ARMA DE FUEGO                 A PIE   
2      ARMA BLANCA / CORTOPUNZANTE                 A PIE   
3                    ARMA DE FUEGO                 A PIE   
4      ARMA BLANCA / CORTOPUNZANTE                 A PIE   
...                            ...                   ...   
11960                ARMA DE FUEGO                 A PIE   
11961  ARMA BLANCA / CORTOPUNZANTE                 A PIE   
11962  ARMA BLANCA / CORTOPUNZANTE                 A PIE   
11963  ARMA BLANCA / CORTOPUNZANTE                 A PIE   
11964  ARMA BLANCA / CORTOPUNZANTE                 A PIE   

              Móvil Victima  Edad       Sexo Estado civil País de nacimiento  \
0      PASAJERO MOTOCICLETA    20  MASCULINO      SOLTERO           COLOMBIA   
1                     A PIE    21  MASCULINO      SOLTERO           COLOMBIA   
2                     A PIE    22  MASCULINO  UNION LIBRE           COLOMBIA   
3                     A PIE    46  MASCULINO      SOLTERO           COLOMBIA   
4                     A PIE    37  MASCULINO  UNION LIBRE           COLOMBIA   
...                     ...   ...        ...          ...                ...   
11960                 A PIE    19  MASCULINO      SOLTERO           COLOMBIA   
11961                 A PIE    30  MASCULINO      SOLTERO           COLOMBIA   
11962                 A PIE    40   FEMENINO      SOLTERO           COLOMBIA   
11963                 A PIE    45  MASCULINO      SOLTERO           COLOMBIA   
11964                 A PIE    54   FEMENINO      SOLTERO           COLOMBIA   

         Clase de empleado Profesión Escolaridad Código DANE  Cantidad  
0      EMPLEADO PARTICULAR         -  SECUNDARIA  91,001,000         1  
1            INDEPENDIENTE         -    PRIMARIA   5,091,000         1  
2            INDEPENDIENTE         -  SECUNDARIA   5,212,000         1  
3            INDEPENDIENTE         -  SECUNDARIA   5,240,000         1  
4            INDEPENDIENTE         -    PRIMARIA   5,250,000         1  
...                    ...       ...         ...         ...       ...  
11960  EMPLEADO PARTICULAR         -  SECUNDARIA  76,001,000         1  
11961          DESEMPLEADO         -    PRIMARIA  76,147,000         1  
11962          AMA DE CASA         -  SECUNDARIA  76,403,000         1  
11

**Nota:** Para reconocer instrucciones de Python vamos a trabajar en este *Notebook* con el conjunto de datos *hospital*

In [ ]:
hospital

EDAD DIABETES HOSPITALIZACIÓN ULTIMO MES    PSA BIOPSIAS PREVIAS  \
0    53.0       NO                         NO   4.00               NO   
1    56.0       NO                         NO   7.70               NO   
2    57.0       NO                         NO  29.00               SI   
3    56.0       NO                         NO   7.00               NO   
4    55.0       NO                         NO  29.00               SI   
..    ...      ...                        ...    ...              ...   
565  69.0       NO                         NO   4.43              NaN   
566  67.0       NO                         NO   8.00               NO   
567  59.0       NO                         NO   9.00               SI   
568   NaN      NaN                        NaN    NaN              NaN   
569   NaN      NaN                        NaN    NaN              NaN   

    VOLUMEN PROSTATICO ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS  \
0                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
1                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
2                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
3                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
4                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
..                 ...                                     ...   
565                NaN          FLUOROQUINOLONA_AMINOGLICOSIDO   
566                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
567                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
568                NaN                                     NaN   
569                NaN                                     NaN   

     NUMERO DE MUESTRAS TOMADAS  CUP ENF. CRONICA PULMONAR OBSTRUCTIVA  \
0                          12.0   NO                                NO   
1                          12.0   NO                                NO   
2                          24.0   NO                                NO   
3                          12.0   NO                                NO   
4                          24.0   NO                                NO   
..                          ...  ...                               ...   
565                        20.0  NaN                                NO   
566                        12.0   NO                               NaN   
567                        12.0   NO                               NaN   
568                         NaN  NaN                               NaN   
569                         NaN  NaN                               NaN   

                       BIOPSIA  \
0                          NEG   
1                          NEG   
2                          NEG   
3                          NEG   
4                          NEG   
..                         ...   
565                        NEG   
566  ADENOCARCINOMA GLEASON 7    
567                        NEG   
568                        NaN   
569                        NaN   

    NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA  \
0                                                    1                          
1                                                    1                          
2                                                    1                          
3                                                    1                          
4                                                    1                          
..                                                 ...                          
565                                                 NO                          
566                                                 NO                          
567                                                 NO                          
568                                                NaN                          
569                                                NaN                          

    FIEBRE  ITU TIPO DE CULTIVO AGENTE AISLADO  \
0       SI   N

# <FONT SIZE=5 COLOR="green"> 3. Exploración de los datos </FONT>

Cuando vamos a explorar los datos. Lo primero que debemos entender es el contexto de los datos.

El conjunto de datos corresponde a: Hospitalización de pacientes por complicaciones infecciosas después de Biopsia Prostática.

1. Antecedentes del paciente.

  - **EDAD**: Edad del paciente

2. Morbilidad asociada al paciente.

  - **DIABETES** Pacience con o sin diabetes
  - **HOSPITALIZACIÓN ÚLTIMO MES**: Indica si el paciente fue hospitalizado el último mes
  - **CUP**: Uso de catéter urinario al momento de la biopsia.
  - **ENF. CRÓNICA PULMONAR OBSTRUCTIVA**: Indica si el paciente tiene una enfermedad crónica pulmonar obstructiva.
  - **VOLUMEN PROSTÁTICO**: Indica si el volumen prostático es mayor a $40 \ cm^3$

3. Antecedentes relacionados con la toma de la biopsia.

  - **PSA**: Concentración de PSA en la sangre (antígeno prostático específico)
  - **BIOPSIAS PREVIAS**: Indica si el paciente a tenido biopsias previas.
  - **ANTIBIOTICO USADO EN LA PROFILAXIS**: Indica el tipo de antibiótico
  - **NÚMERO DE MUESTRAS TOMADAS**: Número de muestras tomadas en la biopsia.
  - **BIOPSIA**: Resultado de la Biopsia

4. Complicaciones infecciosas.

  - **NÚMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA**.
  - **FIEBRE**: Indica si el paciente presenta fiebre.
  - **ITU**: Indica si el paciente presenta infección de tracto urinario.
  - **TIPO DE CULTIVO**: Tipo de cultivo encontrado.
  - **AGENTE AISLADO**: Tipo de agente aislado.
  - **PATRÓN DE RESISTENCIA**: Indica si presenta algún patrón de resistencia.
  - **HOSPITALIZACIÓN**: Indica si el paciente fue hospitalizado luego del procedimiento.
  - **DIAS DE HOSPITALIZACIÓN MQ**: Indica los días de hospitalización médico quirúrgico
  - **DIAS DE HOSPÍTALIZACIÓN UPC**: Indica los días de hospitalización en estado crítico


In [ ]:
# no se recomienda mostrar todos los datos.
hospital

## <FONT SIZE=5 COLOR="green"> 3.1 Instrucciones de exploración general </FONT>

En esta sección revisaremos:

- Los primeros datos
- Los últimos datos
- El tamaño de los datos
- El nombre de las columnas
- Información del conjunto de datos
. Datos faltantes
- Estadísticas generales

1. Usamos la instrucción **head()** para explorar la cabeza de los datos

- Podemos indicar cuántos datos queremos visualizar simplemente con $head(n)$, sonde $n$ indica el número de registros que queremos ver.

In [ ]:
# para visualizar los primeros datos
hospital.head()

EDAD DIABETES HOSPITALIZACIÓN ULTIMO MES   PSA BIOPSIAS PREVIAS  \
0  53.0       NO                         NO   4.0               NO   
1  56.0       NO                         NO   7.7               NO   
2  57.0       NO                         NO  29.0               SI   
3  56.0       NO                         NO   7.0               NO   
4  55.0       NO                         NO  29.0               SI   

  VOLUMEN PROSTATICO ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS  \
0                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
1                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
2                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
3                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
4                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   

   NUMERO DE MUESTRAS TOMADAS CUP ENF. CRONICA PULMONAR OBSTRUCTIVA BIOPSIA  \
0                        12.0  NO                                NO     NEG   
1                        12.0  NO                                NO     NEG   
2                        24.0  NO                                NO     NEG   
3                        12.0  NO                                NO     NEG   
4                        24.0  NO                                NO     NEG   

  NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA  \
0                                                  1                          
1                                                  1                          
2                                                  1                          
3                                                  1                          
4                                                  1                          

  FIEBRE ITU TIPO DE CULTIVO AGENTE AISLADO  \
0     SI  NO              NO             NO   
1     SI  NO              NO             NO   
2     SI  NO     HEMOCULTIVO         E.COLI   
3     SI  NO              NO             NO   
4     SI  NO     HEMOCULTIVO         E.COLI   

                  PATRON DE RESISTENCIA HOSPITALIZACION  \
0                                    NO              SI   
1                                    NO              SI   
2   AMPI R, CIPRO R, GENTA R, SULFA M R              SI   
3                                    NO              SI   
4                                    NO              SI   

   DIAS HOSPITALIZACION MQ  DIAS HOSPITALIZACIÓN UPC  
0                      2.0                       0.0  
1                      5.0                       0.0  
2                      4.0                       3.0  
3                      5.0                       0.0  
4                      4.0                       3.0

2. Usamos la instrucción **tail()** para explorar los últimos datos

In [ ]:
hospital

EDAD DIABETES HOSPITALIZACIÓN ULTIMO MES    PSA BIOPSIAS PREVIAS  \
0    53.0       NO                         NO   4.00               NO   
1    56.0       NO                         NO   7.70               NO   
2    57.0       NO                         NO  29.00               SI   
3    56.0       NO                         NO   7.00               NO   
4    55.0       NO                         NO  29.00               SI   
..    ...      ...                        ...    ...              ...   
565  69.0       NO                         NO   4.43              NaN   
566  67.0       NO                         NO   8.00               NO   
567  59.0       NO                         NO   9.00               SI   
568   NaN      NaN                        NaN    NaN              NaN   
569   NaN      NaN                        NaN    NaN              NaN   

    VOLUMEN PROSTATICO ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS  \
0                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
1                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
2                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
3                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
4                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
..                 ...                                     ...   
565                NaN          FLUOROQUINOLONA_AMINOGLICOSIDO   
566                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
567                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
568                NaN                                     NaN   
569                NaN                                     NaN   

     NUMERO DE MUESTRAS TOMADAS  CUP ENF. CRONICA PULMONAR OBSTRUCTIVA  \
0                          12.0   NO                                NO   
1                          12.0   NO                                NO   
2                          24.0   NO                                NO   
3                          12.0   NO                                NO   
4                          24.0   NO                                NO   
..                          ...  ...                               ...   
565                        20.0  NaN                                NO   
566                        12.0   NO                               NaN   
567                        12.0   NO                               NaN   
568                         NaN  NaN                               NaN   
569                         NaN  NaN                               NaN   

                       BIOPSIA  \
0                          NEG   
1                          NEG   
2                          NEG   
3                          NEG   
4                          NEG   
..                         ...   
565                        NEG   
566  ADENOCARCINOMA GLEASON 7    
567                        NEG   
568                        NaN   
569                        NaN   

    NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA  \
0                                                    1                          
1                                                    1                          
2                                                    1                          
3                                                    1                          
4                                                    1                          
..                                                 ...                          
565                                                 NO                          
566                                                 NO                          
567                                                 NO                          
568                                                NaN                          
569                                                NaN                          

    FIEBRE  ITU TIPO DE CULTIVO AGENTE AISLADO  \
0       SI   N

In [ ]:
# para visualizar los últimos datos
hospital.tail(4)

EDAD DIABETES HOSPITALIZACIÓN ULTIMO MES  PSA BIOPSIAS PREVIAS  \
566  67.0       NO                         NO  8.0               NO   
567  59.0       NO                         NO  9.0               SI   
568   NaN      NaN                        NaN  NaN              NaN   
569   NaN      NaN                        NaN  NaN              NaN   

    VOLUMEN PROSTATICO ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS  \
566                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
567                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
568                NaN                                     NaN   
569                NaN                                     NaN   

     NUMERO DE MUESTRAS TOMADAS  CUP ENF. CRONICA PULMONAR OBSTRUCTIVA  \
566                        12.0   NO                               NaN   
567                        12.0   NO                               NaN   
568                         NaN  NaN                               NaN   
569                         NaN  NaN                               NaN   

                       BIOPSIA  \
566  ADENOCARCINOMA GLEASON 7    
567                        NEG   
568                        NaN   
569                        NaN   

    NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA  \
566                                                 NO                          
567                                                 NO                          
568                                                NaN                          
569                                                NaN                          

    FIEBRE  ITU TIPO DE CULTIVO AGENTE AISLADO PATRON DE RESISTENCIA  \
566     NO   NO              NO            NaN                    NO   
567     NO   NO              NO             NO                    NO   
568    NaN  NaN             NaN            NaN                   NaN   
569    NaN  NaN             NaN            NaN                   NaN   

    HOSPITALIZACION  DIAS HOSPITALIZACION MQ  DIAS HOSPITALIZACIÓN UPC  
566              NO                      0.0                       0.0  
567              NO                      0.0                       0.0  
568             NaN                      NaN                       NaN  
569             NaN                      NaN                       NaN

3. Con la instrucción **shape** miramos el tamaño de nuestros datos

In [ ]:
# para revisar el tamaño de los datos
hospital.shape

(570, 20)

es decir, el conjunto de datos tiene 570 registros o filas y 20 columnas.

4. Para explorar nombre de las variables o columnas **columns**

In [ ]:
# nombres de las columnas
hospital.columns

Index(['EDAD', 'DIABETES', 'HOSPITALIZACIÓN ULTIMO MES', 'PSA',
       'BIOPSIAS PREVIAS', 'VOLUMEN PROSTATICO',
       'ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS', 'NUMERO DE MUESTRAS TOMADAS',
       'CUP', 'ENF. CRONICA PULMONAR OBSTRUCTIVA', 'BIOPSIA',
       'NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA',
       'FIEBRE', 'ITU', 'TIPO DE CULTIVO', 'AGENTE AISLADO',
       'PATRON DE RESISTENCIA', 'HOSPITALIZACION', 'DIAS HOSPITALIZACION MQ',
       'DIAS HOSPITALIZACIÓN UPC'],
      dtype='object')

Podemos ver que tenemos columnas como la edad, diabetes, hospitalización, PSA, etc.

5. **info()** es un resumen de las variables y su clasificación. Además, podemos identificar datos faltantes (NA) en cada una de ellas.

In [ ]:
# resumen de los datos
hospital.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 570 entries, 0 to 569
Data columns (total 20 columns):
 #   Column                                                                     Non-Null Count  Dtype  
---  ------                                                                     --------------  -----  
 0   EDAD                                                                       568 non-null    float64
 1   DIABETES                                                                   568 non-null    object 
 2   HOSPITALIZACIÓN ULTIMO MES                                                 568 non-null    object 
 3   PSA                                                                        564 non-null    float64
 4   BIOPSIAS PREVIAS                                                           566 non-null    object 
 5   VOLUMEN PROSTATICO                                                         567 non-null    object 
 6   ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS                   

Podemos observar:

- Conjunto de datos tiene 570 registros.
- Existen variables que no tienen todos los registros. Es decir, tienen datos faltantes
- Tipo de agente aislado es la que más datos faltantes tiene.


6. Para revisar las estadística generales de las variables **describe()**

- Cuantitativas : ***datos.describe()***

- Cualitativas: ***datos.describe(include='object')***

In [ ]:
# resumen estadístico variables cuantitativas
hospital.describe()

EDAD         PSA  NUMERO DE MUESTRAS TOMADAS  \
count  568.000000  564.000000                  568.000000   
mean    63.882042   13.271401                   12.996479   
std      9.281517   15.668966                    3.921180   
min     39.000000    0.600000                    4.000000   
25%     59.000000    5.600000                   12.000000   
50%     64.000000    8.000000                   12.000000   
75%     69.000000   15.000000                   12.000000   
max    151.000000  100.000000                   24.000000   

       DIAS HOSPITALIZACION MQ  DIAS HOSPITALIZACIÓN UPC  
count               568.000000                568.000000  
mean                  0.165493                  0.014085  
std                   0.958066                  0.196474  
min                   0.000000                  0.000000  
25%                   0.000000                  0.000000  
50%                   0.000000                  0.000000  
75%                   0.000000                  0.000000  
max                  12.000000                  3.000000

In [ ]:
hospital.describe(include = "object")

DIABETES HOSPITALIZACIÓN ULTIMO MES BIOPSIAS PREVIAS  \
count       568                        568              566   
unique        2                          2                2   
top          NO                         NO               NO   
freq        505                        563              510   

       VOLUMEN PROSTATICO ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS  CUP  \
count                 567                                     568  567   
unique                  2                                       5    2   
top                    SI          FLUOROQUINOLONA_AMINOGLICOSIDO   NO   
freq                  408                                     448  563   

       ENF. CRONICA PULMONAR OBSTRUCTIVA BIOPSIA  \
count                                566     568   
unique                                 4      11   
top                                   NO     NEG   
freq                                 546     322   

       NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA  \
count                                                 568                          
unique                                                  6                          
top                                                    NO                          
freq                                                  542                          

       FIEBRE  ITU TIPO DE CULTIVO AGENTE AISLADO PATRON DE RESISTENCIA  \
count     568  568             568            551                   568   
unique      2    2               4              3                     5   
top        NO   NO              NO             NO                    NO   
freq      531  562             560            543                   561   

       HOSPITALIZACION  
count              565  
unique               2  
top                 NO  
freq               541

In [ ]:
# exportar el conjunto de datos hospital a csv
hospital.to_csv("hospital.csv")

In [ ]:
aptos.describe()

Unnamed: 0       precio         mt2     estrato     alcobas  \
count  694.000000   694.000000  694.000000  694.000000  694.000000   
mean   347.500000   317.762739  120.943818    4.651297    2.910663   
std    200.484829   247.614914   74.695493    1.208205    0.763962   
min      1.000000    25.000000   26.000000    2.000000    1.000000   
25%    174.250000   160.000000   71.000000    4.000000    3.000000   
50%    347.500000   245.000000   97.500000    5.000000    3.000000   
75%    520.750000   380.000000  141.750000    6.000000    3.000000   
max    694.000000  1700.000000  500.000000    6.000000   14.000000   

            banos  administracion       avaluo  
count  694.000000      694.000000   694.000000  
mean     2.289625        0.263556   182.636593  
std      0.795718        0.251586   169.082384  
min      1.000000        0.000000     0.149000  
25%      2.000000        0.090000    73.000000  
50%      2.000000        0.201000   131.431500  
75%      3.000000        0.375000   234.051000  
max      6.000000        2.280000  1540.620000

**¿Qué puede decir de la tabla anterior?**

In [ ]:
# include = "object" resumen pequeño de variables cuali
hospital.describe(include='object')

In [ ]:
hospital.columns

Index(['EDAD', 'DIABETES', 'HOSPITALIZACIÓN ULTIMO MES', 'PSA',
       'BIOPSIAS PREVIAS', 'VOLUMEN PROSTATICO',
       'ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS', 'NUMERO DE MUESTRAS TOMADAS',
       'CUP', 'ENF. CRONICA PULMONAR OBSTRUCTIVA', 'BIOPSIA',
       'NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA',
       'FIEBRE', 'ITU', 'TIPO DE CULTIVO', 'AGENTE AISLADO',
       'PATRON DE RESISTENCIA', 'HOSPITALIZACION', 'DIAS HOSPITALIZACION MQ',
       'DIAS HOSPITALIZACIÓN UPC'],
      dtype='object')

In [ ]:
# acceder a una variable
hospital["TIPO DE CULTIVO"]

0               NO
1               NO
2      HEMOCULTIVO
3               NO
4      HEMOCULTIVO
          ...     
565             NO
566             NO
567             NO
568            NaN
569            NaN
Name: TIPO DE CULTIVO, Length: 570, dtype: object

In [ ]:
hospital.HOSPITALIZACION

0       SI
1       SI
2       SI
3       SI
4       SI
      ... 
565     NO
566     NO
567     NO
568    NaN
569    NaN
Name: HOSPITALIZACION, Length: 570, dtype: object

In [ ]:
# las categorias que tiene la variable
hospital["ENF. CRONICA PULMONAR OBSTRUCTIVA"].unique()

array(['NO', 'SI, EPOC', 'SI, ASMA', 'SI', nan], dtype=object)

In [ ]:
hospital["ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS"].unique()

array(['FLUOROQUINOLONA_AMINOGLICOSIDO', 'CEFALOSPORINA_AMINOGLUCOCIDO',
       'OROQUINOLONAS', 'FLUOROQUINOLONA_AMINOGLICÓSIDO', 'OTROS', nan],
      dtype=object)

7. Para acceder a una o varias variable del data.frame podemos usar la instrucción

$$datos[``\text{nombre de la variable}"]$$

$$datos.variable$$

veamos algunos ejemplos


In [ ]:
hospital.columns

Index(['EDAD', 'DIABETES', 'HOSPITALIZACIÓN ULTIMO MES', 'PSA',
       'BIOPSIAS PREVIAS', 'VOLUMEN PROSTATICO',
       'ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS', 'NUMERO DE MUESTRAS TOMADAS',
       'CUP', 'ENF. CRONICA PULMONAR OBSTRUCTIVA', 'BIOPSIA',
       'NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA',
       'FIEBRE', 'ITU', 'TIPO DE CULTIVO', 'AGENTE AISLADO',
       'PATRON DE RESISTENCIA', 'HOSPITALIZACION', 'DIAS HOSPITALIZACION MQ',
       'DIAS HOSPITALIZACIÓN UPC'],
      dtype='object')

In [ ]:
hospital["ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS"].value_counts()

ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS
FLUOROQUINOLONA_AMINOGLICOSIDO    448
CEFALOSPORINA_AMINOGLUCOCIDO       56
OTROS                              42
OROQUINOLONAS                      18
FLUOROQUINOLONA_AMINOGLICÓSIDO      4
Name: count, dtype: int64

In [ ]:
# seleccionar la variable precio
hospital["EDAD"]

Si queremos más variables podemos usar:

In [ ]:
hospital[["EDAD", "DIABETES"]]

## <FONT SIZE=5 COLOR="green"> 3.2 Tabla de frecuencias </FONT>

Para hacer el conteo de las categorías o clases de una variables podemos usar la instrucción:

$$datos[``variable"].value\_counts()$$

Ejemplo: Para hacer una tabla de frecuencias de la variable *HOSPITALIZACIÓN*

In [ ]:
# tabla de frecuencia
hospital["HOSPITALIZACION"].value_counts()

In [ ]:
aptos.head()

Unnamed: 0  precio    mt2 ubicacion  estrato  alcobas  banos balcon  \
0           1    79.0  43.16     norte        3        3      1     si   
1           2    93.0  56.92     norte        2        2      1     si   
2           3   100.0  66.40     norte        3        2      2     no   
3           4   123.0  61.85     norte        2        3      2     si   
4           5   135.0  89.80     norte        4        3      2     si   

  parqueadero  administracion     avaluo terminado  
0          si           0.050  14.923002        no  
1          si           0.069  27.000000        si  
2          no           0.000  15.738427        no  
3          si           0.130  27.000000        no  
4          no           0.000  39.567000        si

In [ ]:
# Ejercicio: hacer la tabla de frecuencia de la variable ENF. CRONICA PULMONAR OBSTRUCTIVA
hospital["ENF. CRONICA PULMONAR OBSTRUCTIVA"].value_counts()

## <FONT SIZE=5 COLOR="green"> 3.3 Tabla de frecuencias cruzadas </FONT>

Para hacer una tabla de frecuencias cruzadas entre dos o más variables usamos la instrucción:

$$pd.crosstab(datos[``variable1"], datos[``variable1"])$$

In [ ]:
# hacer la tabla cruzada de HOSPITALIZACIÓN Y DIABETES
pd.crosstab(hospital["ENF. CRONICA PULMONAR OBSTRUCTIVA"] , hospital["FIEBRE"])

FIEBRE                              NO  SI
ENF. CRONICA PULMONAR OBSTRUCTIVA         
NO                                 509  37
SI                                  12   0
SI, ASMA                             4   0
SI, EPOC                             4   0

También se pueden relacionar dos o más variables para la *crosstable*

In [ ]:
pd.crosstab([hospital["HOSPITALIZACION"], hospital["FIEBRE"]]  , hospital["DIABETES"])

# <FONT SIZE=5 COLOR="green"> 4. Manejo de DataFrames </FONT>

En esta sección revisaremos algunas operaciones y acciones que son importantes de conocer en un *dataframe*:

- Ordenar por una variable

- Renombrar una variable

- Eliminar una variable

- Agregar una variable

- Agrupar por categorias de una variable y aplicar funciones.

Para repasar

https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf

## <FONT SIZE=4 COLOR="magenta"> 4.1 Ordenar según una variable </FONT>

Para ordenar por una variable usamos la instrucción

$$datos.sort\_values(variable, ascending = False) $$

se usa:

- *ascending=True* :  ordena de menor a mayor.
- *ascending=False* : ordena de mayor a menor.


In [ ]:
hospital

EDAD DIABETES HOSPITALIZACIÓN ULTIMO MES    PSA BIOPSIAS PREVIAS  \
0    53.0       NO                         NO   4.00               NO   
1    56.0       NO                         NO   7.70               NO   
2    57.0       NO                         NO  29.00               SI   
3    56.0       NO                         NO   7.00               NO   
4    55.0       NO                         NO  29.00               SI   
..    ...      ...                        ...    ...              ...   
565  69.0       NO                         NO   4.43              NaN   
566  67.0       NO                         NO   8.00               NO   
567  59.0       NO                         NO   9.00               SI   
568   NaN      NaN                        NaN    NaN              NaN   
569   NaN      NaN                        NaN    NaN              NaN   

    VOLUMEN PROSTATICO ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS  \
0                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
1                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
2                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
3                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
4                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
..                 ...                                     ...   
565                NaN          FLUOROQUINOLONA_AMINOGLICOSIDO   
566                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
567                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
568                NaN                                     NaN   
569                NaN                                     NaN   

     NUMERO DE MUESTRAS TOMADAS  CUP ENF. CRONICA PULMONAR OBSTRUCTIVA  \
0                          12.0   NO                                NO   
1                          12.0   NO                                NO   
2                          24.0   NO                                NO   
3                          12.0   NO                                NO   
4                          24.0   NO                                NO   
..                          ...  ...                               ...   
565                        20.0  NaN                                NO   
566                        12.0   NO                               NaN   
567                        12.0   NO                               NaN   
568                         NaN  NaN                               NaN   
569                         NaN  NaN                               NaN   

                       BIOPSIA  \
0                          NEG   
1                          NEG   
2                          NEG   
3                          NEG   
4                          NEG   
..                         ...   
565                        NEG   
566  ADENOCARCINOMA GLEASON 7    
567                        NEG   
568                        NaN   
569                        NaN   

    NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA  \
0                                                    1                          
1                                                    1                          
2                                                    1                          
3                                                    1                          
4                                                    1                          
..                                                 ...                          
565                                                 NO                          
566                                                 NO                          
567                                                 NO                          
568                                                NaN                          
569                                                NaN                          

    FIEBRE  ITU TIPO DE CULTIVO AGENTE AISLADO  \
0       SI   N

In [ ]:
# ordenamos por la variable Edad de menor a mayor





EDAD DIABETES HOSPITALIZACIÓN ULTIMO MES   PSA BIOPSIAS PREVIAS  \
181  151.0       NO                         NO   5.7               NO   
161  143.0       SI                         NO   6.8               NO   
46    84.0       NO                         NO  38.9               NO   
48    84.0       NO                         NO  40.0               NO   
531   84.0       NO                         NO  78.0               NO   
..     ...      ...                        ...   ...              ...   
414   42.0       NO                         NO   6.0               NO   
481   39.0       NO                         NO   3.9               NO   
224   39.0       NO                         NO   3.6               NO   
568    NaN      NaN                        NaN   NaN              NaN   
569    NaN      NaN                        NaN   NaN              NaN   

    VOLUMEN PROSTATICO ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS  \
181                 SI                                   OTROS   
161                 NO          FLUOROQUINOLONA_AMINOGLICOSIDO   
46                  SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
48                  SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
531                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
..                 ...                                     ...   
414                 NO            CEFALOSPORINA_AMINOGLUCOCIDO   
481                 NO            CEFALOSPORINA_AMINOGLUCOCIDO   
224                 NO            CEFALOSPORINA_AMINOGLUCOCIDO   
568                NaN                                     NaN   
569                NaN                                     NaN   

     NUMERO DE MUESTRAS TOMADAS  CUP ENF. CRONICA PULMONAR OBSTRUCTIVA  \
181                        12.0   NO                                NO   
161                        12.0   NO                                NO   
46                          6.0   SI                                NO   
48                          6.0   SI                                NO   
531                        10.0   NO                                NO   
..                          ...  ...                               ...   
414                        12.0   NO                                NO   
481                        12.0   NO                                NO   
224                        12.0   NO                                NO   
568                         NaN  NaN                               NaN   
569                         NaN  NaN                               NaN   

                        BIOPSIA  \
181                         NEG   
161                         NEG   
46   ADENOCARCINOMA GLEASON 10    
48   ADENOCARCINOMA GLEASON 10    
531   ADENOCARCINOMA GLEASON 8    
..                          ...   
414                         NEG   
481                         NEG   
224                         NEG   
568                         NaN   
569                         NaN   

    NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA  \
181                                                 NO                          
161                                                 NO                          
46                                                  NO                          
48                                                  NO                          
531                                                 NO                          
..                                                 ...                          
414                                                 NO                          
481                                                 NO                          
224                                                 NO                          
568                                                NaN                          
569                                                NaN                          

    FIEBRE  ITU TIPO DE CULTIVO AGENTE AISLADO PATRON

## <FONT SIZE=4 COLOR="magenta"> 4.2 Renombrar una variable </FONT>

Podemos renombrar una columna de la siguiente forma usando la instrucción ***rename***

$$ datos.rename(columns=\{key : value, key : value\}, inplace=True) $$

- Para cambiar el nombre de una variable debemos usar un diccionario donde:
    
  - ***key*** son los nombres actuales de las variables.
  - ***value*** : son los nombres nuevos de las variables.

- La instrucción **inplace** se usa para determinar si cambiamos el *dataframe* original.

In [ ]:
hospital.columns

Index(['EDAD', 'DIABETES', 'HOSPITALIZACIÓN ULTIMO MES', 'PSA',
       'BIOPSIAS PREVIAS', 'VOLUMEN PROSTATICO',
       'ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS', 'NUMERO DE MUESTRAS TOMADAS',
       'CUP', 'ENF. CRONICA PULMONAR OBSTRUCTIVA', 'BIOPSIA',
       'NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA',
       'FIEBRE', 'ITU', 'TIPO DE CULTIVO', 'AGENTE AISLADO',
       'PATRON DE RESISTENCIA', 'HOSPITALIZACION', 'DIAS HOSPITALIZACION MQ',
       'DIAS HOSPITALIZACIÓN UPC'],
      dtype='object')

In [ ]:
hospital.rename(columns = {"ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS" : "Anti" } , inplace = True )
# inplace = True ; para que sobreescriba y tenga efecto en adelante.

In [ ]:
hospital.head()

In [ ]:
# vamos a cambiar el nombre largo de la variable NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA por algo más corto
hospital.rename(columns = { "NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA" : "INF_POST_BIOPSIA"  } ).head(5)

In [ ]:
# para que aplique en adelante
hospital.rename(columns = { "NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA" : "INF_POST_BIOPSIA"  } , inplace=True)

In [ ]:
hospital.head()

In [ ]:
# eliminar una variable diabetes
hospital.drop("DIABETES", axis = 1 , inplace = True)

In [ ]:
hospital.head()

EDAD HOSPITALIZACIÓN ULTIMO MES   PSA BIOPSIAS PREVIAS VOLUMEN PROSTATICO  \
0  53.0                         NO   4.0               NO                 SI   
1  56.0                         NO   7.7               NO                 SI   
2  57.0                         NO  29.0               SI                 SI   
3  56.0                         NO   7.0               NO                 SI   
4  55.0                         NO  29.0               SI                 SI   

                             Anti  NUMERO DE MUESTRAS TOMADAS CUP  \
0  FLUOROQUINOLONA_AMINOGLICOSIDO                        12.0  NO   
1  FLUOROQUINOLONA_AMINOGLICOSIDO                        12.0  NO   
2  FLUOROQUINOLONA_AMINOGLICOSIDO                        24.0  NO   
3  FLUOROQUINOLONA_AMINOGLICOSIDO                        12.0  NO   
4  FLUOROQUINOLONA_AMINOGLICOSIDO                        24.0  NO   

  ENF. CRONICA PULMONAR OBSTRUCTIVA BIOPSIA  \
0                                NO     NEG   
1                                NO     NEG   
2                                NO     NEG   
3                                NO     NEG   
4                                NO     NEG   

  NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA  \
0                                                  1                          
1                                                  1                          
2                                                  1                          
3                                                  1                          
4                                                  1                          

  FIEBRE ITU TIPO DE CULTIVO AGENTE AISLADO  \
0     SI  NO              NO             NO   
1     SI  NO              NO             NO   
2     SI  NO     HEMOCULTIVO         E.COLI   
3     SI  NO              NO             NO   
4     SI  NO     HEMOCULTIVO         E.COLI   

                  PATRON DE RESISTENCIA HOSPITALIZACION  \
0                                    NO              SI   
1                                    NO              SI   
2   AMPI R, CIPRO R, GENTA R, SULFA M R              SI   
3                                    NO              SI   
4                                    NO              SI   

   DIAS HOSPITALIZACION MQ  DIAS HOSPITALIZACIÓN UPC  
0                      2.0                       0.0  
1                      5.0                       0.0  
2                      4.0                       3.0  
3                      5.0                       0.0  
4                      4.0                       3.0

## <FONT SIZE=4 COLOR="magenta"> 4.3 Eliminar filas y columnas</FONT>

Para eliminar una columna o una fila usamos la instrucción ***.drop***

$$datos.drop(``nombre", axis = 1, inplace = T \, or \, F)$$

- *axis = 1* : es por columna
- *axis = 0* : es por  fila
- *inplace = T* : sobreescribe en el *dataframe* original.
- *inplace = F* : no sobreescribe. Mantiene el original

In [ ]:
# eliminar la variable DIAS HOSPITALIZACION MQ
hospital.drop("DIAS HOSPITALIZACION MQ", axis = 1 , inplace = True)

In [ ]:
hospital.head(5)

In [ ]:
# Ejercicio: Eliminar las variables : DIAS HOSPITALIZACIÓN UPC ,  ANTIBIOTICO UTILIZADO EN LA PROFILAXIS


## <FONT SIZE=4 COLOR="magenta"> 4.4 Agrupaciones: ***groupby()*** </FONT>

- Se usa para agrupar una variable por sus categorías y luego aplicar una función a estos grupos con respecto a otra variable

<center><img src="https://github.com/Fabian830348/cursos/blob/master/intro/groupby1.png?raw=true" alt="centered image" width="450" height="350"></center> <center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: Elaboración propia  </FONT> <figcaption></center>


Podemos hacer operaciones con las agrupaciones, así:

$$datos.groupby(``Variable")[``\text{segunda variable}"].funcion()$$

In [ ]:
# aplicación de group_by
hospital.groupby("HOSPITALIZACION")

**¿Qué podemos hacer con esto?**

In [ ]:
hospital.groupby("HOSPITALIZACION").EDAD.describe()

# <FONT SIZE=5 COLOR="green"> 5. Algunas funciones importantes </FONT>



1. Media o promedio: **.mean()**

- Es el "equilibrio" de los datos.
- Se usa para variables cuantitativas.
- Siempre está entre el mínimo y máximo de los datos.
- Es sensible a datos atípicos. Por está razón se debe tener cuidado con su interpretación.

In [ ]:
hospital.EDAD.mean()

2. Mediana: **.median()**

- Es el valor que divide los datos en dos partes iguales (50%-50%)
- Aplica para variables cuantitativas
- Puede o no estar en el conjunto de datos. Depende si se tiene un número par o impar de elementos.
- No se ve afectada por la existencia de datos atípicos. Por está razón algunas veces es usada para imputar valores atípicos.

In [ ]:
hospital.EDAD.median()

3. Moda: **.mode()**

- Es el valor que más se repite. Es decir el que tiene mayor frecuencia
- Aplica para variables cualitativas y cuantitativas
- UN conjunto puede no tener o tener múltiples modas.

In [ ]:
hospital.EDAD.mode()

4. Máximo y mínimo del conjunto: **.max** y **.min**

In [ ]:
hospital.EDAD.max()

In [ ]:
hospital.EDAD.min()

# <FONT SIZE=5 COLOR="green"> 6. Eliminar los NA's </FONT>

Para eliminar los NA´s podemos usar la instrucción

$$datos.dropna(inplace=True)$$

Toca tener especial cuidado en eliminar NA porque se puede reducir el conjunto de datos considerablemente.

In [ ]:
hospital.dropna(inplace=False)

In [ ]:
hospital.info()

In [ ]:
hospital.head(10)

Observe que cuando se elimina una fila el índice también se elimina. En el ejemplo anterior no aparece el índice 9. Para arreglar esta situación usamo la instrucción

$$hospital.reset\_index$$

In [ ]:
hospital.reset_index(inplace=True)

In [ ]:
hospital.info()

# <FONT SIZE=5 COLOR="green"> 7. Algunas gráficas importantes </FONT>

En esta sección revisaremos algunas gráficas clásicas.

Para esto podemos usar diferentes librerías: *mathplotlib* y *plotly*. También se pueden hacer gráficos elaborados en *seaborn* Más documentación la pueden encontrar en los siguientes links.

- https://matplotlib.org/

- https://seaborn.pydata.org/

- https://plotly.com/python/plotly-express/

**Nota:** nos centraremos más en la parte gráfica y el código que en el análisis de las gráficas.

<center><FONT SIZE=5 COLOR="Magenta"> Mathplotlib - Seaborn - Plotly.Express </FONT></center>

Existen múltiples librerías para hacer gráficas en Python. En esta parte mostraremos algunos ejemplos, sin embargo, invitamos al lector a explorar con más detalle estas librerías.

- **Mathplotlib**: es una biblioteca de Python utilizada para trazar gráficos con la ayuda de otras bibliotecas como Numpy y Pandas. Es una poderosa herramienta para visualizar datos en Python. Se utiliza para crear inferencias estáticas y trazar gráficos 2D de matrices. Fue presentado por primera vez por John D. Hunter en 2002.


- **Seaborn**: también es una biblioteca de Python que se utiliza para trazar gráficos con la ayuda de Matplotlib, Pandas y Numpy. Está construido sobre el techo de Matplotlib y se considera como un superconjunto de la biblioteca Matplotlib. Ayuda a visualizar datos univariados y bivariados.

- **Plotly.express** es otra librería para visualización con gráficos interactivos que muestran algunos detalles de las gráficas.

Vamos a cargar las librerías

In [ ]:
import matplotlib.pyplot as plt
import plotly.express as px
import seaborn as sb

## <FONT SIZE=4 COLOR="green"> 7.1. TABLA DE FRECUENCIA - BARRAS Y TORTAS </FONT>

Para hacer el gráfico de barras o de tortas haremos primero una tabla de frecuencias.

- Recordemos de los gráficos de barras o de tortas se usan en general para variables categóricas o para variables cuantitativas con pocos valores, como es el caso de *número de hijos*.

- En el eje $x$ van las categorías

- En el eje $y$ la cantidad o frecuencia de cada clase o categoria

- En los diagramas de barras las barras van separadas y de colocan de diferente color.

In [ ]:
px.bar(hospital.HOSPITALIZACION.value_counts(),
       color = ["No" , "Si"])

In [ ]:
# A partir de la tabla de frecuencias de la variable hospitalización.
freq = hospital.HOSPITALIZACION.value_counts()
px.pie(freq,                               # tabla de frecuencias
       values= freq.values,                # valores de las frecuencias absolutas
       names= freq.index)                  # nombres de las categorias


## <FONT SIZE=4 COLOR="green"> 7.2 HISTOGRAMA </FONT>

- El histograma se usa para variables de tipo cuantitativo continuo o discreto con muchas mediciones diferentes (como la edad en ciertos casos).

- En el eje $x$ se identifican las clases (intervalos) de la variable.

- En el eje $y$ se identifica la frecuencia de cada una de las clases.

- Las barras en el histograma quedan contiguas y además se utiliza un solo color en estas.

In [ ]:
# Histograma más básico
px.histogram(hospital.EDAD)

Podemos colocarle más elementos de edición

In [ ]:
fig = px.histogram(hospital,                               # variable
                   nbins = 30,                             # número de intervalos
                   x = "EDAD",                             # nombre de la variable x
                   title = "Histograma con edición"        # título del gráfico
                                      )

fig.update_traces(marker_color = 'cyan',                    # color de relleno
                  marker_line_color = 'red',                # color del borde
                  marker_line_width = 1,                    # grueso de la línea
                  opacity = 1)                              # transparencia del color

fig.show()

## <FONT SIZE=4 COLOR="green"> 7.3. BOXPLOT (GRÁFICO DE CAJAS Y BIGOTES) </FONT>

- El gráfico de caja, boxplot del inglés, fue ideado por Jhon Tukey en 1977 y es una representación estadística que resalta principalmente aspectos de la distribución de los datos u observaciones por medio de las medidas de posición.

- Los diagramas de caja brindan información valiosa que no se observa en un histograma como son los cuartiles.

A continuación a describir cada uno de los elementos que tiene un boxplot

**La caja rectangular** : el gráfico contiene una caja rectangular. Es importante indicar que esta caja no tiene coordenadas cartesianas en el eje $X$  y las alturas en el eje $y$ no representa una escala de frecuencias. Las líneas horizontales en el diagrama de cajas describen el primer y tercer cuartil. $Q_1$ y $Q_3$ respectivamente.

**Mediana** : es una línea que se muestra dentro de la caja rectangular y hace referencia al valor que divide los datos en dos partes iguales. Recuerde que la mediana es igual a $Q_2$.

**Bigotes** : son líneas que salen a los costados de la caja y se usan como referencia para ubicar las observaciones que estan por fuera del $50 \%$ central de distribucion. Los bigotes nos dan información sobre la existencia de valores atípicos (*outliers*).

In [ ]:
# boxplot clásico
px.box(hospital.EDAD)

Podemos comenzar a hacer gráficos más eleborados que relacionen más variables

In [ ]:
px.box(hospital,                                                # nombre de los datos
       x = "HOSPITALIZACION",                                    # variable cualitativa o categórica
       y = "EDAD",                                              # variable cuantitativa
       title = "Boxplot variable EDAD vs HOSPITALIZACIÓN",      # título del gráfico
       color = "HOSPITALIZACION"                                # color por la variable categórica
       )

In [ ]:
px.box(hospital,                                                                # nombre de los datos
       y = "EDAD",                                                              # variable cuantitativa
       x = "ENF. CRONICA PULMONAR OBSTRUCTIVA",                                 # variable cualitativa o categórica
       title = "Boxplot variable EDAD vs ENF. CRONICA PULMONAR OBSTRUCTIVA",    # título del gráfico
       color = "ENF. CRONICA PULMONAR OBSTRUCTIVA"                              # color por la variable categórica
        )

## <FONT SIZE=4 COLOR="green"> 7.4 DIAGRAMA DE DISPERSIÓN </FONT>

- Los diagramas de dispersión permiten representar graficamente dos variables cuantitativas.

In [ ]:
px.scatter(hospital,
           x= "EDAD",
           y= "PSA")

**Ejercicio**

Aplicar el siguiente código al conjunto Apartamento.csv.

In [ ]:
pip install sweetviz

In [ ]:
import sweetviz as sw

In [ ]:
reporte=sw.analyze(aptos, target_feat="precio")

In [ ]:
from IPython.display import HTML
reporte.show_html()
HTML("SWEETVIZ_REPORT.html")

# <FONT SIZE=5 COLOR="green"> 8. Adicional: Acceder a los elementos del DataFrame </FONT>

Para acceder a los elementos, o a las filas o columnas del DataFrame podemos usar:

- ***iloc*** : para acceder por la posición
- ***loc.*** : para acceder por el nombre

Si usamos la primera instrucción

- ***df.iloc[i,j]*** : elemento de la fila i y columna j.

- ***df.iloc[i]*** : devuelve la fila i

- ***df.iloc[lista i , lista j]*** : devuelve un data.frame con las filas en lista i y columnas en columna j.

- ***df.iloc[:, j]*** : devuelve toda columna j.

Si usamos la instrucción **.loc**

- ***df.loc[nombre fila , nombre columna]***: es lo mismo anterior pero con los nombres.

- ***df[nombre columna]*** : devuelve las columnas en name columnas

- ***df.columna*** : otra forma para acceder a una columna que en el nombre no tenga espacios en blanco.

## <FONT SIZE=4 COLOR="magenta"> 8.1 Ejemplos con **.iloc** </FONT>




In [ ]:
aptos.head()


<center><img src="https://github.com/Fabian830348/cursos/blob/master/intro/iloc1.png?raw=true" alt="centered image" width="600" height="350"></center> <center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: Elaboración propia  </FONT> <figcaption></center>

In [ ]:
# para seleccionar un elemento de la fila i, columna j.
aptos.iloc[1,3]

In [ ]:
# para seleccionar una fila
aptos.iloc[3]

In [ ]:
# para seleccionar una columna
aptos.iloc[ : , 1]

<center><img src="https://github.com/Fabian830348/cursos/blob/master/intro/iloc2.png?raw=true" alt="centered image" width="450" height="350"></center> <center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: Elaboración propia  </FONT> <figcaption></center>

In [ ]:
# filas 1 y 3 y columnas 0 y 1
aptos.iloc[[1,3], [0,1]]

<center><img src="https://github.com/Fabian830348/cursos/blob/master/intro/iloc3.png?raw=true" alt="centered image" width="450" height="350"></center> <center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: Elaboración propia  </FONT> <figcaption></center>

In [ ]:
# filas de la 1 a la 3 , tener en cuenta que termina un índice antes.
aptos.iloc[1:4]

<center><img src="https://github.com/Fabian830348/cursos/blob/master/intro/iloc4.png?raw=true" alt="centered image" width="450" height="350"></center> <center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: Elaboración propia  </FONT> <figcaption></center>

In [ ]:
# columnas 1 en adelante, ojo los dos puntos significan : end
# : solos toman todas las filas
# 1: de 1 en adelante
aptos.iloc[:, 1:]

Veamos algunos ejemplos adicionales.

In [ ]:
# las columnas 2-5
aptos.iloc[: , 2:6]

In [ ]:
# para seleccionar la segunda, tercera y primera fila
aptos.iloc[[3,4,0]]

In [ ]:
# para seleccionar la segunda y cuarta columna
aptos.iloc[: , [1,3]]

## <FONT SIZE=5 COLOR="magenta">8.2 Ejemplos con **loc** </FONT>

Veamos algunos ejemplos con **.loc**.

Recordemos que en *.iloc* se busca por posición y en *.loc* por nombre.

In [ ]:
# veamos el nombre de las columnas
aptos.columns

In [ ]:
# veamos el nombre de las filas
aptos.index

In [ ]:
# selecciono la fila 0 y 1
aptos.loc[[0,1]]

In [ ]:
aptos.loc[: , "precio"]